docs/cardano-monads.rst

```diff
@@ -0,0 +1,608 @@
+introduce local effects, there's a high cost in boilerplate. This can be already
+observed in our codebase.
+
+Point-by-point rundown:
+
+* Flexibility: LOW. Extension of environment requires declaring a new type,
+  writing ``Has`` instances for it, and potentially monad instances (although the
+  approach recommends avoiding monad classes in favor of ``Has`` classes, not
+  everything is under our control and there are classes by external libraries).
+  There's no way to partially run effects -- one needs to supply the entire
+  ``ReaderT`` environment up front.
+
+* Extensibility: MODERATE. Defining new effects boils down to declaring a method
+  record and a ``Has`` class for it. Then, for each field in the method record we
+  must define a function that extracts this method using the ``Has`` class and
+  runs it. Boilerplate-heavy, yes, but at least effects don't need to know about
+  each other.
+
+* Ease of use: MODERATE. Beginners can quickly grasp the basic concepts, but
+  there are pitfalls when it comes to lifting method records (see section on
+  ``SendActions``).
+
+* Compile-time performance: SUPERB. There's nothing to it, GHC can easily handle
+  this style of code.
+
+* Run-time performance: MODERATE. A flat ``ReaderT`` layer is good, but method
+  records are stored within the context (rather than passed via instance
+  search). Because of this, the compiler can't assume that the method record is
+  coherent (it's not, you can modify it with ``local``), and it limits inlining.
+  Instance specialization can't be performed because there are no instances to
+  speak of.
+
+* Run-time configurability: SUPERB. You can put just about anything into these
+  manually defined method records, modify them at will, etc. However, we need to
+  figure out a good story for method records that depend on other method
+  records, because it's the same pitfall as with ``SendActions``: how can we
+  ensure that if we change the something (say, a logging method) with ``local``,
+  other method records that use logging will be updated accordingly?
+
+* Predictability: LOW/HIGH. Depends on whether we have a ``newtype`` over
+  ``ReaderT``.
+
+Verdict: the approach is viable but has its costs.
+
+
+Dictionary-Passing Style
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to previous approach and one was widely used by the networking team.
+
+As we understand ideas behind it:
+
+- Single base monad
+
+  - Similarly to ``ReaderT IO`` approach
+
+- Differentiating implementations all stored in records
+
+  - No type classes and instances like ``instance MyClass (ReaderT Ctx IO)``
+  - Each function accepts a single **explicit** parameter of type ``Ctx m``
+  - This type ``Ctx`` contains all needed methods
+
+Point-by-point rundown:
+
+* Flexibility: LOW. (Same as above).
+
+* Extensibility: GOOD. (Same as above, minus a bit of boilerplate helpers).
+
+* Ease of use: LOW. (Same as above, plus one needs to manually pass a the
+  context everywhere).
```

Comment:
This is a bit too harsh in my opinion. The context doesn't need to be passed *everywhere*, but only in the select few places where it's required. That is in many cases an advantage over throwing it into a reader context: with the reader, the dictionary is passed along to *every* term, even those who don't need it, whereas with explicit dictionary passing, terms which use fewer effects have fewer function arguments and so can be used with fewer obligations / are more flexible.

Reply:
> The context doesn't need to be passed everywhere

I think there's a misunderstanding, because the article discusses monadic effects in general (which are ubiquitous), while `SendActions` is just an example of the manual dictionary passing approach.

Nearly all code in cardano-sl requires some sort of monadic context, so I feel that the use of "everywhere" is justified here.

Reply:
> The context doesn't need to be passed everywhere, but only in the select few places where it's required. That is in many cases an advantage over throwing it into a reader context: with the reader, the dictionary is passed along to every term, even those who don't need it, whereas with explicit dictionary passing, terms which use fewer effects have fewer function arguments and so can be used with fewer obligations / are more flexible.

I don't get it. If your function doesn't need any context, just make it pure without any reader context and then use `pure` where it's called. If your function requires only some part of the whole context, change constraint type to `(MonadReader ctx m, HasSmallContext ctx)`. Then the function will use fewer effects and can be used with fewer obligations / be more flexible.

Reply:
> I don't get it. If your function doesn't need any context, just make it pure without any reader context and then use pure where it's called.

Sure, that's the way to do it. And if your function does need context, but only a small part of it?

> If your function requires only some part of the whole context, change constraint type to (MonadReader ctx m, HasSmallContext ctx). Then the function will use fewer effects and can be used with fewer obligations / be more flexible.

I argue it's more flexible to just take the obligations as a function argument. Instead of `f :: (MonadRreader ctx m, HasSmallContext ctx) => m t`, it would be `g :: SmallContext -> t`. The unnecessary parameter `m` goes away and we have a simpler type. This is what the `MonadReader ctx m` solution boils down to: to define `f` you'd end up defining `g` anyway, so why not just define `g` and leave it at that?